sqlglotrs/src/tokenizer.rs

```diff
@@ -0,0 +1,642 @@
+use crate::settings::TokenizerSettings;
+use crate::token::{Token, TokenType};
+use crate::trie::{Trie, TrieResult};
+use pyo3::prelude::*;
+
+#[derive(Debug)]
+#[pyclass]
+pub struct Tokenizer {
+    settings: TokenizerSettings,
+    keyword_trie: Trie,
+}
+
+#[pymethods]
+impl Tokenizer {
+    #[new]
+    #[pyo3(signature = (settings=None))]
+    pub fn new(settings: Option<TokenizerSettings>) -> Tokenizer {
+        let settings = settings.unwrap_or(TokenizerSettings::default());
+        let mut keyword_trie = Trie::new();
+        let single_token_strs: Vec<String> = settings
+            .single_tokens
+            .keys()
+            .map(|s| s.to_string())
+            .collect();
+        let trie_filter =
+            |key: &&String| key.contains(" ") || single_token_strs.iter().any(|t| key.contains(t));
+
+        keyword_trie.add(settings.keywords.keys().filter(trie_filter));
+        keyword_trie.add(settings.comments.keys().filter(trie_filter));
+        keyword_trie.add(settings.quotes.keys().filter(trie_filter));
+        keyword_trie.add(settings.format_strings.keys().filter(trie_filter));
+
+        Tokenizer {
+            settings,
+            keyword_trie,
+        }
+    }
+
+    pub fn tokenize(&self, sql: &str) -> Vec<Token> {
+        let mut state = TokenizerState::new(sql, &self.settings, &self.keyword_trie);
+        state.tokenize()
+    }
+}
+
+#[derive(Debug)]
+struct TokenizerState<'a> {
+    sql: Vec<char>,
+    size: usize,
+    tokens: Vec<Token>,
+    start: usize,
+    current: usize,
+    line: usize,
+    column: usize,
+    comments: Vec<String>,
+    is_end: bool,
+    current_char: char,
+    peek_char: char,
+    previous_token_line: Option<usize>,
+    keyword_trie: &'a Trie,
+    settings: &'a TokenizerSettings,
+}
+
+impl<'a> TokenizerState<'a> {
+    fn new(
+        sql: &str,
+        settings: &'a TokenizerSettings,
+        keyword_trie: &'a Trie,
+    ) -> TokenizerState<'a> {
+        TokenizerState {
+            sql: sql.chars().collect(),
+            size: sql.len(),
+            tokens: Vec::new(),
+            start: 0,
+            current: 0,
+            line: 1,
+            column: 0,
+            comments: Vec::new(),
+            is_end: false,
+            current_char: '\0',
+            peek_char: '\0',
+            previous_token_line: None,
+            keyword_trie,
+            settings,
+        }
+    }
+
+    fn tokenize(&mut self) -> Vec<Token> {
+        self.scan(None);
+        std::mem::replace(&mut self.tokens, Vec::new())
```

Comment:
what is this

Reply:
Instead of copying the vector we're moving it here, but since we can leave the variable uninitialized we replace it with the empty vector.